col*: add tree.Datum-backed vector for unoptimized types

yuzefovich opened the pull request:
**colmem: reset flat bytes vector when reused**

Our `MaybeAppendColumn` reuses the same vector if a vector of the
desired type is already in the requested position. However, the flat
bytes vector needs special treatment when reused - it needs to be reset
and previously we forgot to do so. This is now fixed.

Release note (bug fix): Previously, in some cases an internal error
could occur when queries that have columns of BYTES type in the output
were executed via the vectorized engine, and this has been fixed.

**coldata: introduce ColumnFactory interface**

Upcoming introduction of datum-backed support of all types that are not
explicitly supported by the vectorized engine would introduce
a dependency on `tree` package from users of `coldata` other than
`colexec` (mainly I'm thinking about `workload`). We don't want to have
that, and in order to prevent it this commit introduces
`coldata.ColumnFactory` interface which is responsible for making
columns. All types that have explicity support in `coldata` will be
created by `coldata.StandartColumnFactory` and the datum-backed ones
will be created by an extended column factory which will be introduced
in the follow-up work.

Release note: None

**coldataext: add tree.Datum-backed vector for unoptimized types**

This commit introduces `tree.Datum`-backed vector that is used as the
fallback for all unoptimized types (i.e. those that we don't have an
explicit implementation for). As a result, all types are now supported
by the vectorized engine. Comparison and hash overloads have been added,
so we can now do things like joins and distinct on these datum-backed
types, but no binary nor comparison (other than the "common six")
operators are supported at the moment. Additionally, no comparisons
between "native" and datum-backed types are supported, and neither are
casts of datum-backed types.

`coldata.DatumVec` interface has been introduced with the implementation
living in `coldataext` - this is done to avoid importing `tree` package
into `coldata`. An extended column factory is added in `coldataext` as
well, and the vectorized engine has been transitioned onto using the
extended column factory. Additionally, all `tree.Datum`s are wrapped
into `coldataext.Datum` so that we can define necessary methods on it
(like `CompareDatum` and `Hash`).

The canonical type family for all datum-backed types is the special
`typeconv.DatumVecCanonicalTypeFamily` value.

Fixes: #45612.

Release note (sql change): Vectorized execution engine now supports
(although not "natively" - meaning we don't have optimized
representations for the newly-supported types) all data types that
CockroachDB supports.

**sqlsmith: remove restriction on vectorizable types**

Release note: None

**colexec: miscellaneous cleanups**

This commit addresses a couple of TODOs I left while adding datum-backed
vector as well as cleans up a few things. Namely, it does the following:
- merges the case of Date and Oid families in key encoding with Int
family since they have the same physical representation
- cleans up boolVecToSelOp to expect that it operates only on boolean
column by making a separate planning utility that'll plan a zero
operator for unknown family
- `ConstNullOp` has been clean up a bit (now it simply sets the nulls
vector to "all nulls", without paying attention to the selection vector)
- now that we handle `types.UnknownFamily` well, `execplan.go` has been
cleaned up to remove "plan maybe null" stuff.

Release note: None

**colexec, colflow: add type hydration**

Recent addition of enum types started requiring special "type hydration"
before the types are safe to use by the execution engine. In the row
execution, it is done during every processor's initialization; however,
the operators don't call the same `ProcessorBase.Init` method, so we
need to do the hydration ourselves.

This commit adds the hydration to `colBatchScan` (because it gets the
types from the table descriptor and is special in this way) and to
`setupInput` when setting up the vectorized flow. All operators (apart
from `colBatchScan`) get the types from `InputSyncSpec`, so `setupInput`
is the convenient place to do the hydration to make all operators happy.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/48422)
<!-- Reviewable:end -->

yuzefovich committed:
coldata: introduce ColumnFactory interface

Upcoming introduction of datum-backed support of all types that are not
explicitly supported by the vectorized engine would introduce
a dependency on `tree` package from users of `coldata` other than
`colexec` (mainly I'm thinking about `workload`). We don't want to have
that, and in order to prevent it this commit introduces
`coldata.ColumnFactory` interface which is responsible for making
columns. All types that have explicity support in `coldata` will be
created by `coldata.StandartColumnFactory` and the datum-backed ones
will be created by an extended column factory which will be introduced
in the follow-up work.

Release note: None

yuzefovich committed:
coldataext: add tree.Datum-backed vector for unoptimized types

This commit introduces `tree.Datum`-backed vector that is used as the
fallback for all unoptimized types (i.e. those that we don't have an
explicit implementation for). As a result, all types are now supported
by the vectorized engine. Comparison and hash overloads have been added,
so we can now do things like joins and distinct on these datum-backed
types, but no binary nor comparison (other than the "common six")
operators are supported at the moment. Additionally, no comparisons
between "native" and datum-backed types are supported, and neither are
casts of datum-backed types.

`coldata.DatumVec` interface has been introduced with the implementation
living in `coldataext` - this is done to avoid importing `tree` package
into `coldata`. An extended column factory is added in `coldataext` as
well, and the vectorized engine has been transitioned onto using the
extended column factory. Additionally, all `tree.Datum`s are wrapped
into `coldataext.Datum` so that we can define necessary methods on it
(like `CompareDatum` and `Hash`).

The canonical type family for all datum-backed types is the special
`typeconv.DatumVecCanonicalTypeFamily` value.

Release note (sql change): Vectorized execution engine now supports
(although not "natively" - meaning we don't have optimized
representations for the newly-supported types) all data types that
CockroachDB supports.

yuzefovich committed:
sqlsmith: remove restriction on vectorizable types

Release note: None

yuzefovich committed:
colexec: miscellaneous cleanups

This commit addresses a couple of TODOs I left while adding datum-backed
vector as well as cleans up a few things. Namely, it does the following:
- merges the case of Date and Oid families in key encoding with Int
family since they have the same physical representation
- cleans up boolVecToSelOp to expect that it operates only on boolean
column by making a separate planning utility that'll plan a zero
operator for unknown family
- `ConstNullOp` has been clean up a bit (now it simply sets the nulls
vector to "all nulls", without paying attention to the selection vector)
- now that we handle `types.UnknownFamily` well, `execplan.go` has been
cleaned up to remove "plan maybe null" stuff.

Release note: None

asubiotto reviewed (changes requested):
> Reviewed 41 of 41 files at r1, 80 of 80 files at r2, 2 of 2 files at r3, 7 of 7 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48422)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/col/coldataext/datum_vec.go, line 11 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M78KmITBo44IX2wc_h8:-M78KmITBo44IX2wc_h9:b-7yt7o1) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c7622adb5cfd82a9b2b8313f1113536d4ada4a2/pkg/col/coldataext/datum_vec.go#L11)):*
> ```Go
> // licenses/APL.txt.
> 
> package coldataext
> ```

I think it'd be nice to name this something more specific. What about something like `coldatum`?
___
*[pkg/col/coldataext/datum_vec.go, line 30 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M78LO0nAHbOWRW4PMxp:-M78LO0nAHbOWRW4PMxq:bdzz2u2) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c7622adb5cfd82a9b2b8313f1113536d4ada4a2/pkg/col/coldataext/datum_vec.go#L30)):*
> ```Go
> type DatumWithEvalCtx struct {
> 	tree.Datum
> 	evalCtx *tree.EvalContext
> ```

What's the reason to store an evalCtx with each datum? It seems expensive.
___
*[pkg/col/coldataext/datum_vec.go, line 64 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M78LJHo8DG8MfTMP7Xi:-M78LJHo8DG8MfTMP7Xj:b-54ygy0) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c7622adb5cfd82a9b2b8313f1113536d4ada4a2/pkg/col/coldataext/datum_vec.go#L64)):*
> ```Go
> // so that we do not overload tree.Datum.Compare method.
> func (d *DatumWithEvalCtx) CompareDatum(other interface{}) int {
> 	return d.Datum.Compare(d.evalCtx, maybeUnwrapDatumWithEvalCtx(other))
> ```

Not sure I understand the need for `maybeUnwrapDatumWithEvalCtx` in this case
___
*[pkg/col/coldataext/extended_column_factory.go, line 39 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M79-yW7EDYK2LQWDl3q:-M79-yW7EDYK2LQWDl3r:b-ftai4r) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c7622adb5cfd82a9b2b8313f1113536d4ada4a2/pkg/col/coldataext/extended_column_factory.go#L39)):*
> ```Go
> 		return newDatumVec(t, n, cf.evalCtx)
> 	}
> 	return coldata.StandardColumnFactory.MakeColumn(t, n)
> ```

Do we use the standard column factory anywhere? Might be better to just have a single factory implementation if not.
___
*[pkg/col/colserde/record_batch.go, line 36 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M78NDw525C0yieZFz6x:-M78NDw6-kkpNQSJiqyA:bfp3cil) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c7622adb5cfd82a9b2b8313f1113536d4ada4a2/pkg/col/colserde/record_batch.go#L36)):*
> ```Go
> // the given type.
> func numBuffersForType(t *types.T) int {
> 	// Most types are represented by 3 memory.Buffers. One buffer for the
> ```

I would add it's because they're serialized using flat bytes.
___
*[pkg/sql/colexec/case_test.go, line 61 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M79112vCjqXydRBHuHl:-M79112vCjqXydRBHuHm:bwj6881) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b52063ec66ee1fb883eeec4f7523d571394b371/pkg/sql/colexec/case_test.go#L61)):*
> ```Go
> 			// Test the short-circuiting behavior.
> 			tuples:     tuples{{1, 2}, {2, 0}, {nil, nil}, {3, 3}},
> 			renderExpr: "CASE WHEN @1 = 2 THEN 0::FLOAT WHEN @1 / @2 = 1 THEN 1::FLOAT END",
> ```

OOC why the change to FLOAT?
___
*[pkg/sql/sqlbase/column_type_encoding.go, line 891 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M791RjV7BPjAReLqPoA:-M791RjV7BPjAReLqPoB:b-46r6p) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c7622adb5cfd82a9b2b8313f1113536d4ada4a2/pkg/sql/sqlbase/column_type_encoding.go#L891)):*
> ```Go
> 		}
> 		return tree.NewDJSON(jsonDatum), nil
> 	// TODO(yuzefovich): is it intentional that types.TupleFamily is missing?
> ```

I don't know, we should figure it out.
___
*[pkg/sql/sqlbase/column_type_encoding.go, line 1052 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M79-cLl3cghltfO0sr0:-M79-cLl3cghltfO0sr1:b5jz396) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c7622adb5cfd82a9b2b8313f1113536d4ada4a2/pkg/sql/sqlbase/column_type_encoding.go#L1052)):*
> ```Go
> // DecodeArrayNoMarshalColumnValue skips the step where the MarshalColumnValue
> // is stripped from the bytes. This is required for single-column family arrays.
> func DecodeArrayNoMarshalColumnValue(
> ```

I might be missing something, but why are these public now? I don't think I see this used in this commit.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48422)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/col/coldataext/datum_vec.go, line 11 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M78KmITBo44IX2wc_h8:-M79de-q33jd6KWceuPV:bw7prbd) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c7622adb5cfd82a9b2b8313f1113536d4ada4a2/pkg/col/coldataext/datum_vec.go#L11)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I think it'd be nice to name this something more specific. What about something like `coldatum`?
</blockquote></details>

I'm thinking of `coldataext` as an "extension of `coldata`". `coldatum` might be a little confusing since "data" is the plural form of "datum", and we would have `coldata` and `coldatum`. I dunno. To me `coldataext` makes more sense - like use `coldata` if you want the "standard" version and use `coldataext" if you want the "extended" version.
___
*[pkg/col/coldataext/datum_vec.go, line 30 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M78LO0nAHbOWRW4PMxp:-M79cbvg9-07369LGKSo:bhkwn7u) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c7622adb5cfd82a9b2b8313f1113536d4ada4a2/pkg/col/coldataext/datum_vec.go#L30)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

What's the reason to store an evalCtx with each datum? It seems expensive.
</blockquote></details>

This is a way to plumb the eval context into the generated code. If you look at `CompareDatum` you'll see that we need the eval context to come from somewhere, and there are a few options:
1. add it as an argument to `CompareDatum` - this will require us to plumb eval context into all operators that use datum-backed vector in their generated code, so I don't like it that much
2. get it from `datumVec` somehow - this will require either passing it as an argument (which seems wrong) or storing the pointer to `datumVec` inside of `DatumWithEvalCtx`. The latter option would be pretty much the same as storing the eval context itself.

I feel like this extra pointer should not be a big concern in the grand scheme of things - I still expect that the interface conversions will be taking a lot more time and the `tree.Datum`s themselves a lot more space than having this `*tree.EvalContext`.
___
*[pkg/col/coldataext/datum_vec.go, line 64 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M78LJHo8DG8MfTMP7Xi:-M79gBmIBlTjADNxz6oL:b-wmxgsb) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c7622adb5cfd82a9b2b8313f1113536d4ada4a2/pkg/col/coldataext/datum_vec.go#L64)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Not sure I understand the need for `maybeUnwrapDatumWithEvalCtx` in this case
</blockquote></details>

Here `other` could be two things:
1. `other == nil`, and we need to update to `tree.DNull` (this is because we store nulls in a separate `Nulls` vector, so `datumVec.data[j]` could be left unset when we set a NULL value)
2. `other` is `DatumWithEvalCtx`, so we need to get `tree.Datum` out of it.

The context is that we have something like this:
```
var dv1, dv2 datumVec
dv1.Get(i).CompareDatum(dv2.Get(j))
```
`maybeUnwrapDatumWithEvalCtx` has an additional case for when `other` is `tree.Datum` which shouldn't actually occur here, but I think it's ok to reuse that function.
___
*[pkg/col/coldataext/extended_column_factory.go, line 39 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M79-yW7EDYK2LQWDl3q:-M79jYnW1vNgywaZrSOn:bg9e5ai) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c7622adb5cfd82a9b2b8313f1113536d4ada4a2/pkg/col/coldataext/extended_column_factory.go#L39)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Do we use the standard column factory anywhere? Might be better to just have a single factory implementation if not.
</blockquote></details>

There are two users of `coldata` at this moment:
1. the vectorized engine (i.e. `pkg/sql/colexec` and its dependencies), this will use the extended column factory
2. the workload (i.e. `pkg/workload`), and this will use the standard factory.

The whole dance with the `ColumnFactory` interface is done so that we don't bring the dependency on `sem/tree` and some other packages (that support `tree.Datum`) into `workload` package. In some sense, I view `coldata` as the package that we will be putting "native" support for the types, and packages like `workload` that don't need the datum-backed vector should only have access to those "native" types.
___
*[pkg/col/colserde/record_batch.go, line 36 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M78NDw525C0yieZFz6x:-M79kutGCFJqyZTa0az2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c7622adb5cfd82a9b2b8313f1113536d4ada4a2/pkg/col/colserde/record_batch.go#L36)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I would add it's because they're serialized using flat bytes.
</blockquote></details>

Done.
___
*[pkg/sql/colexec/case_test.go, line 61 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M79112vCjqXydRBHuHl:-M79l1reDTVg2calvZro:buz9il1) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b52063ec66ee1fb883eeec4f7523d571394b371/pkg/sql/colexec/case_test.go#L61)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

OOC why the change to FLOAT?
</blockquote></details>

That was done because of the limitation that we prohibit casts between non-identical decimals (i.e. rounding is prohibited). For some reason one side we had `width=0, precision=0` and on the other `width=11, precision=2`, I didn't look too closely into this since we might lift that restriction now that we no longer lose type information.
___
*[pkg/sql/sqlbase/column_type_encoding.go, line 891 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M791RjV7BPjAReLqPoA:-M79m5wU0oLwU4lL4s_r:b-pvpzfp) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c7622adb5cfd82a9b2b8313f1113536d4ada4a2/pkg/sql/sqlbase/column_type_encoding.go#L891)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I don't know, we should figure it out.
</blockquote></details>

Note that this is in `sqlbase`, so it is a limitation of the row-by-row engine as well. I think this PR should not concern itself with this fix, but I agree we should understand why tuple family is omitted.
___
*[pkg/sql/sqlbase/column_type_encoding.go, line 1052 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M79-cLl3cghltfO0sr0:-M79lmOG4JLw-0Y0WAcv:b-6nrhqe) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c7622adb5cfd82a9b2b8313f1113536d4ada4a2/pkg/sql/sqlbase/column_type_encoding.go#L1052)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I might be missing something, but why are these public now? I don't think I see this used in this commit.
</blockquote></details>

Nice catch, this is a refactor artifact, fixed.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
@mjibson could you please take a look at the third commit?

asubiotto reviewed (changes requested):
> Reviewed 86 of 86 files at r5, 3 of 79 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48422)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

___
*[pkg/col/coldataext/datum_vec.go, line 30 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M78LO0nAHbOWRW4PMxp:-M7CBE6r1-8vPn1J-Qo_:b-sj7mw4) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c7622adb5cfd82a9b2b8313f1113536d4ada4a2/pkg/col/coldataext/datum_vec.go#L30)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

This is a way to plumb the eval context into the generated code. If you look at `CompareDatum` you'll see that we need the eval context to come from somewhere, and there are a few options:
1. add it as an argument to `CompareDatum` - this will require us to plumb eval context into all operators that use datum-backed vector in their generated code, so I don't like it that much
2. get it from `datumVec` somehow - this will require either passing it as an argument (which seems wrong) or storing the pointer to `datumVec` inside of `DatumWithEvalCtx`. The latter option would be pretty much the same as storing the eval context itself.

I feel like this extra pointer should not be a big concern in the grand scheme of things - I still expect that the interface conversions will be taking a lot more time and the `tree.Datum`s themselves a lot more space than having this `*tree.EvalContext`.
</blockquote></details>

For option 1, instead of plumbing it into the operator itself, won't we have a datum vector that we can extract it from?
___
*[pkg/col/typeconv/typeconv.go, line 47 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M7CAeNc5SQkaTzfczss:-M7CAeNd73mOuf6G2ASP:b-aeav31) ([raw file](https://github.com/cockroachdb/cockroach/blob/d78dcf26204b8b0f96dfadec339e17b65be7d2f8/pkg/col/typeconv/typeconv.go#L47)):*
> ```Go
> 		return types.IntervalFamily
> 	default:
> 		// TODO(yuzefovich): consider adding explicit support for
> ```

nit: truncated comment


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48422)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/col/coldataext/datum_vec.go, line 30 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M78LO0nAHbOWRW4PMxp:-M7E48nM39jTZKmWm7Vd:b-7msfw7) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c7622adb5cfd82a9b2b8313f1113536d4ada4a2/pkg/col/coldataext/datum_vec.go#L30)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

For option 1, instead of plumbing it into the operator itself, won't we have a datum vector that we can extract it from?
</blockquote></details>

We do have the datum vectors in scope where the templated call to `CompareDatum` is made, but in order to get the eval context out of it, we will need to change our `compareFunc` from `overloads_base.go` to take in at least one other parameter that is the name of the local variable that stores the vector of the left side.

For example, here is the template we have
```
		arg1 := _L_UNSAFEGET(col1, i)
		arg2 := _R_UNSAFEGET(col2, i)
		_ASSIGN(projCol[i], arg1, arg2)
```
and this is what gets generated:
```
					arg1 := col1.Get(i)
					arg2 := col2.Get(i)

					{
						var cmpResult int

						cmpResult = arg1.(*coldataext.DatumWithEvalCtx).CompareDatum(arg2)

						projCol[i] = cmpResult == 0
					}
```
We would need to modify `_ASSIGN` to take in `col1` (which is `DatumVec`). A similar change will need to be made for `assignFunc` as well, and all templates and template generators will need to be updated.

I think that at this point storing an extra pointer on every datum should not be a big concern. If we do some benchmarks and find that it is a concern, we can revisit this later. Left a TODO for this.
___
*[pkg/col/typeconv/typeconv.go, line 47 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M7CAeNc5SQkaTzfczss:-M7E3ljgEPhPwPaa8wor:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d78dcf26204b8b0f96dfadec339e17b65be7d2f8/pkg/col/typeconv/typeconv.go#L47)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: truncated comment
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

asubiotto commented:
LGTM. Maybe @jordanlewis will want to take a look

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48422)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/col/coldataext/datum_vec.go, line 30 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M78LO0nAHbOWRW4PMxp:-M7Nh7Ub0xWN3w4HkCVo:b-kj3kai) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c7622adb5cfd82a9b2b8313f1113536d4ada4a2/pkg/col/coldataext/datum_vec.go#L30)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

We do have the datum vectors in scope where the templated call to `CompareDatum` is made, but in order to get the eval context out of it, we will need to change our `compareFunc` from `overloads_base.go` to take in at least one other parameter that is the name of the local variable that stores the vector of the left side.

For example, here is the template we have
```
		arg1 := _L_UNSAFEGET(col1, i)
		arg2 := _R_UNSAFEGET(col2, i)
		_ASSIGN(projCol[i], arg1, arg2)
```
and this is what gets generated:
```
					arg1 := col1.Get(i)
					arg2 := col2.Get(i)

					{
						var cmpResult int

						cmpResult = arg1.(*coldataext.DatumWithEvalCtx).CompareDatum(arg2)

						projCol[i] = cmpResult == 0
					}
```
We would need to modify `_ASSIGN` to take in `col1` (which is `DatumVec`). A similar change will need to be made for `assignFunc` as well, and all templates and template generators will need to be updated.

I think that at this point storing an extra pointer on every datum should not be a big concern. If we do some benchmarks and find that it is a concern, we can revisit this later. Left a TODO for this.
</blockquote></details>

I agree with Alfonso that this is objectionable. Benchmarks aside, it's not good practice to store an extra pointer along with every datum in the system. That will bloat memory by 2x.

I don't think that changing _ASSIGN to include one of the columns is that bad, I think it should be included.

I would recommend making a new PR that comes first, which changes _ASSIGN to include the column but do nothing with it. Then, you can edit this PR to remove the DatumWithEvalCtx.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Alright, you guys win, opened a PR to plumb column names as arguments :)

yuzefovich commented:
I added another commit that fixes a bug with flat bytes vector not being reset in `MaybeAppendColumn` function which was exposed due to the last commit (that allowed "untyped nulls" in `execplan.go`). I'll another PR with that commit, and it should go in first.

jordanlewis reviewed (approved):
Exciting! Did you do any performance tests over large amounts of data, before and after? I would like to see some simple ones if you can.

> Reviewed 25 of 41 files at r8, 84 of 84 files at r12.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-:-M7f3PWt760LZ6KzIzAj:b-6o191x)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @yuzefovich)

___
*[pkg/col/coldata/datum_vec.go, line 42 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M7f-qmiBXRh4KHWuJZD:-M7f-qmiBXRh4KHWuJZE:b-1koou9) ([raw file](https://github.com/cockroachdb/cockroach/blob/360937608a387f371fe65bb3d64a749e3141050a/pkg/col/coldata/datum_vec.go#L42)):*
> ```Go
> 	// Len returns the length of the vector.
> 	Len() int
> 	// Cap returns the underlying capacity of the vector.
> ```

Much of this interface seems identical to the Bytes one. Should we pull out the declaration elsewhere?
___
*[pkg/col/coldataext/datum_vec.go, line 76 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M7f0Cg84yrw2F4j4CfL:-M7f0Cg84yrw2F4j4CfM:bbl5q5l) ([raw file](https://github.com/cockroachdb/cockroach/blob/360937608a387f371fe65bb3d64a749e3141050a/pkg/col/coldataext/datum_vec.go#L76)):*
> ```Go
> 
> // Get returns *Datum which wraps the ith tree.Datum in datumVec
> // and the evalCtx stored inside datumVec.
> ```

All of these could just say "Get implements the DatumVec interface"
___
*[pkg/col/colserde/arrowbatchconverter.go, line 142 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M7f12GO4-r1eqRE7uoJ:-M7f12GO4-r1eqRE7uoK:b-flfc0a) ([raw file](https://github.com/cockroachdb/cockroach/blob/360937608a387f371fe65bb3d64a749e3141050a/pkg/col/colserde/arrowbatchconverter.go#L142)):*
> ```Go
> 			}
> 			data = c.builders.binaryBuilder.NewBinaryArray().Data()
> 		case types.AnyFamily:
> ```

AnyFamily? Should this be the `default` case instead? I feel like this might be wrong
___
*[pkg/col/typeconv/typeconv.go, line 49 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M7f1F7s0mOMd7eN1KEl:-M7f1F7s0mOMd7eN1KEm:b-8zk2c7) ([raw file](https://github.com/cockroachdb/cockroach/blob/360937608a387f371fe65bb3d64a749e3141050a/pkg/col/typeconv/typeconv.go#L49)):*
> ```Go
> 		// TODO(yuzefovich): consider adding native support for
> 		// types.UnknownFamily.
> 		return types.AnyFamily
> ```

Hmm, I don't particularly like this. I think it might get confusing, because I believe `types.AnyFamily` has a separate meaning. Ideally, we could make a new value just for internal use, like `genericFamily` or something, and return that... maybe it doesn't matter though? I just don't like using `AnyFamily` as it says in the documentation that `AnyFamily` should not be used at execution time, and we're now violating that.
___
*[pkg/sql/colexec/bool_vec_to_sel.go, line 166 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M7f1wvi3u6Rr3ilb-Yo:-M7f1wvi3u6Rr3ilb-Yp:b-i3hx7i) ([raw file](https://github.com/cockroachdb/cockroach/blob/360937608a387f371fe65bb3d64a749e3141050a/pkg/sql/colexec/bool_vec_to_sel.go#L166)):*
> ```Go
> 			}
> 		}
> 	case types.AnyFamily:
> ```

Why would this operator need to be able to accept non boolean? I don't understand
___
*[pkg/sql/colexec/execgen/cmd/execgen/overloads_base.go, line 891 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M7f2bCI45AnsUALjwkK:-M7f2bCI45AnsUALjwkL:b-de1b0r) ([raw file](https://github.com/cockroachdb/cockroach/blob/360937608a387f371fe65bb3d64a749e3141050a/pkg/sql/colexec/execgen/cmd/execgen/overloads_base.go#L891)):*
> ```Go
> 		// This is somewhat unfortunate, but we can neither use coldata.Datum
> 		// nor tree.Datum because we have generated files living in two
> 		// different packages (sql/colexec and col/coldata).
> ```

I thought we fixed this with coldataext?
___
*[pkg/sql/sqlbase/column_type_encoding.go, line 891 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M7f2vRl6OAWNIxLzg6J:-M7f2vRl6OAWNIxLzg6K:by7aiqw) ([raw file](https://github.com/cockroachdb/cockroach/blob/360937608a387f371fe65bb3d64a749e3141050a/pkg/sql/sqlbase/column_type_encoding.go#L891)):*
> ```Go
> 		}
> 		return tree.NewDJSON(jsonDatum), nil
> 	// TODO(yuzefovich): is it intentional that types.TupleFamily is missing?
> ```

We don't have an encoding for tuple types, you can't store them on disk.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
I haven't done any benchmarks, and I'm not sure what are the interesting examples. Any suggestions? I could choose `TimeTZ` as a representative type and then try some projections and maybe hash join?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-:-M7hs-C5DhoFp0mLVoTD:bs8osru)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/col/coldata/datum_vec.go, line 42 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M7f-qmiBXRh4KHWuJZD:-M7ho0fMC8MGT75YCrAg:b-c221er) ([raw file](https://github.com/cockroachdb/cockroach/blob/360937608a387f371fe65bb3d64a749e3141050a/pkg/col/coldata/datum_vec.go#L42)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Much of this interface seems identical to the Bytes one. Should we pull out the declaration elsewhere?
</blockquote></details>

Yes, the methods are the same in its purpose, but here we operate on untyped `interface{}` and do the conversion at runtime whereas `Bytes` vector has well-typed arguments. I don't think we can unify this without hurting the performance of flat bytes, can we?
___
*[pkg/col/coldataext/datum_vec.go, line 30 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M78LO0nAHbOWRW4PMxp:-M7hoLPl1gM3WXPXHfBG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c7622adb5cfd82a9b2b8313f1113536d4ada4a2/pkg/col/coldataext/datum_vec.go#L30)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I agree with Alfonso that this is objectionable. Benchmarks aside, it's not good practice to store an extra pointer along with every datum in the system. That will bloat memory by 2x.

I don't think that changing _ASSIGN to include one of the columns is that bad, I think it should be included.

I would recommend making a new PR that comes first, which changes _ASSIGN to include the column but do nothing with it. Then, you can edit this PR to remove the DatumWithEvalCtx.
</blockquote></details>

Done.
___
*[pkg/col/coldataext/datum_vec.go, line 76 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M7f0Cg84yrw2F4j4CfL:-M7hodFp8VN67kipn0CP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/360937608a387f371fe65bb3d64a749e3141050a/pkg/col/coldataext/datum_vec.go#L76)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

All of these could just say "Get implements the DatumVec interface"
</blockquote></details>

Done.
___
*[pkg/col/colserde/arrowbatchconverter.go, line 142 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M7f12GO4-r1eqRE7uoJ:-M7hrMEm8Xi-3k8g8lkp:b7e9ffl) ([raw file](https://github.com/cockroachdb/cockroach/blob/360937608a387f371fe65bb3d64a749e3141050a/pkg/col/colserde/arrowbatchconverter.go#L142)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

AnyFamily? Should this be the `default` case instead? I feel like this might be wrong
</blockquote></details>

This can't be `default` because types like `Int2` and `Float` are handled below. We could restructure the code though.

However, I think we need to have a "representative canonical type family" for the datum-backed types, and we should use `case` on that. At the moment, `types.AnyFamily` is that representative. I responded more in depth to your comment about this choice.
___
*[pkg/col/typeconv/typeconv.go, line 49 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M7f1F7s0mOMd7eN1KEl:-M7hp1aIDPIo4d8sPHXr:bt7czhq) ([raw file](https://github.com/cockroachdb/cockroach/blob/360937608a387f371fe65bb3d64a749e3141050a/pkg/col/typeconv/typeconv.go#L49)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Hmm, I don't particularly like this. I think it might get confusing, because I believe `types.AnyFamily` has a separate meaning. Ideally, we could make a new value just for internal use, like `genericFamily` or something, and return that... maybe it doesn't matter though? I just don't like using `AnyFamily` as it says in the documentation that `AnyFamily` should not be used at execution time, and we're now violating that.
</blockquote></details>

I agree that `types.AnyFamily` was designed with a different use case in mind, but I think our use case is somewhat similar and this PR overloads the meaning of `types.AnyFamily`.

We could introduce another family for us, something like `types.DatumBackedCanonicalFamily` to be used only for the purposes of overload resolution in the vectorized engine. Do you think it would be worth it?
___
*[pkg/sql/colexec/bool_vec_to_sel.go, line 166 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M7f1wvi3u6Rr3ilb-Yo:-M7hpjh026d-_niPWHCD:bw0xt87) ([raw file](https://github.com/cockroachdb/cockroach/blob/360937608a387f371fe65bb3d64a749e3141050a/pkg/sql/colexec/bool_vec_to_sel.go#L166)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Why would this operator need to be able to accept non boolean? I don't understand
</blockquote></details>

I fixed it in the fourth commit.

The reason for this change was that we now support `types.UnknownFamily` which is represented as a vector of NULL values only, and there were test failures because of the baked in assumption that the vector is `Bool` but it could be `Datum` with all nils.
___
*[pkg/sql/colexec/execgen/cmd/execgen/overloads_base.go, line 891 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M7f2bCI45AnsUALjwkK:-M7hqSPIC4xaukawmkFr:b-2gkjl1) ([raw file](https://github.com/cockroachdb/cockroach/blob/360937608a387f371fe65bb3d64a749e3141050a/pkg/sql/colexec/execgen/cmd/execgen/overloads_base.go#L891)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I thought we fixed this with coldataext?
</blockquote></details>

The comment refers to inability to use `coldata.Datum` here (because we have `vec_tmpl.go` which generates `col/coldata/vec.eg.go`, so we cannot refer to `coldata.Datum` in there; similarly, we cannot just use `Datum` because `colexec` files would not know what that is). We also cannot move `coldata.Datum` into `coldataext` because that would create an import cycle.

It's not a big deal because `coldata.Datum` is `interface{}` anyway, but it'd be nice to use the former, but I don't see a way to do that.
___
*[pkg/sql/sqlbase/column_type_encoding.go, line 891 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M7f2vRl6OAWNIxLzg6J:-M7hrINb0v0IEpiKQN-p:b-onbugi) ([raw file](https://github.com/cockroachdb/cockroach/blob/360937608a387f371fe65bb3d64a749e3141050a/pkg/sql/sqlbase/column_type_encoding.go#L891)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

We don't have an encoding for tuple types, you can't store them on disk.
</blockquote></details>

I see, removed the TODO.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
I addressed `types.AnyFamily` issue and did some quick benchmarks, and the results are quite nice.

On a single node local cluster I have one table as
```
root@:26257/defaultdb> show create table tz;
  table_name |               create_statement
-------------+------------------------------------------------
  tz         | CREATE TABLE tz (
             |     a INT8 NOT NULL,
             |     t TIMETZ NULL,
             |     CONSTRAINT "primary" PRIMARY KEY (a ASC),
             |     FAMILY "primary" (a, t)
             | )
(1 row)
```
that is populated with 1 million of rows.

For the following query:
```
root@:26257/defaultdb> explain (vec) select count(*) from tz where t < '0000-01-01 20:09:01.255355+00:00'::timetz;
                  text
-----------------------------------------
  │
  └ Node 1
    └ *colexec.countOp
      └ *colexec.selLTDatumDatumConstOp
        └ *colexec.colBatchScan
```
it takes about `1.15s` with `vectorize=off` and `0.94s` with `vectorize=on`. Granted, some speed up comes from the more efficient count operator, but most of it comes from the more efficient filter that has datum comparison.

When doing a hash join on the datum-backed column:
```
root@:26257/defaultdb> explain (vec) select * from tz as t1, tz as t2 where t1.t = t2.t limit 1;
              text
---------------------------------
  │
  └ Node 1
    └ *colexec.limitOp
      └ *colexec.hashJoiner
        ├ *colexec.colBatchScan
        └ *colexec.colBatchScan
```
we also see nice speed up - `1.6s` with `off` vs `0.6s` with `on`. Obviously, here the speed up comes from the significantly faster hash joiner, but the hashing of datums doesn't slow it down.

I'm quite happy with the results on these quick benchmarks.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-:-M7iZTtmAjVM-quP7SsC:bqw0qh9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/col/typeconv/typeconv.go, line 49 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M7f1F7s0mOMd7eN1KEl:-M7iZE25DrGZFPzMYASQ:b-876ot6) ([raw file](https://github.com/cockroachdb/cockroach/blob/360937608a387f371fe65bb3d64a749e3141050a/pkg/col/typeconv/typeconv.go#L49)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I agree that `types.AnyFamily` was designed with a different use case in mind, but I think our use case is somewhat similar and this PR overloads the meaning of `types.AnyFamily`.

We could introduce another family for us, something like `types.DatumBackedCanonicalFamily` to be used only for the purposes of overload resolution in the vectorized engine. Do you think it would be worth it?
</blockquote></details>

I added a special `typeconv.DatumVecCanonicalTypeFamily` for this purpose, and it works like a charm. Thanks for the suggestion.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
colexec, colflow: add type hydration

Recent addition of enum types started requiring special "type hydration"
before the types are safe to use by the execution engine. In the row
execution, it is done during every processor's initialization; however,
the operators don't call the same `ProcessorBase.Init` method, so we
need to do the hydration ourselves.

This commit adds the hydration to `colBatchScan` (because it gets the
types from the table descriptor and is special in this way) and to
`setupInput` when setting up the vectorized flow. All operators (apart
from `colBatchScan`) get the types from `InputSyncSpec`, so `setupInput`
is the convenient place to do the hydration to make all operators happy.

Release note: None

yuzefovich commented:
I added another commit to fix the type hydration issues that appeared after a recent merge of enum work. @rohany could take a look at the fifth commit please?

rohany commented:
Oh this is great! Does this mean that the enum tests will start to go through the vectorized engine now? 

The changes look good, just please add a reference to this issue (https://github.com/cockroachdb/cockroach/issues/49261) for tracking to ensure that those calls eventually look at cached copies of descriptors. Also, i think it could be better to share a single evalContext through those calls to setup input rather than making a new one every time (but I don't think its a big deal if it isn't easy). This makes it so that if we cache any data in the eval context itself about any type information (which I think we will), then we can reuse that across all of these operators.

yuzefovich commented:
Thanks for taking a look!

> Oh this is great! Does this mean that the enum tests will start to go through the vectorized engine now?

Yeah, the vectorized engine will now be able to read the enums as well as do comparisons and hashing of them.

>  just please add a reference to this issue

Done.

> share a single evalContext

Done.

rohany commented:
Final commit lgtm, but I don't think you added the issue reference.

yuzefovich commented:
> but I don't think you added the issue reference

Oh, I misunderstood - I modified the issue itself only.

rohany commented:
Oh ok, never mind!

jordanlewis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, awesome

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-:-M7oY5zT6pcTzcE3QEXx:b-wvb32a)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @yuzefovich)

___
*[pkg/col/coldata/datum_vec.go, line 42 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/48422#-M7f-qmiBXRh4KHWuJZD:-M7oXg11ABnamWPn0JJ7:b-3bppdi) ([raw file](https://github.com/cockroachdb/cockroach/blob/360937608a387f371fe65bb3d64a749e3141050a/pkg/col/coldata/datum_vec.go#L42)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Yes, the methods are the same in its purpose, but here we operate on untyped `interface{}` and do the conversion at runtime whereas `Bytes` vector has well-typed arguments. I don't think we can unify this without hurting the performance of flat bytes, can we?
</blockquote></details>

No, but at least there would be an interface (unused I guess) that would be implemented by both. I guess it isn't really worth the work though, since it won't really be used.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Thanks for the reviews!

bors r+
